validate clip_norm argument in clip_by_norm API

SuryanarayanaY opened the pull request:
The API `clip_by_norm` have argument `clip_norm` which accepts  `0-D (scalar) `Tensor` > 0` . But if we pass `-ve` value for this argument then its not raising intended error and converting the input tensor into Negative which IMO is wrong. Hence I am adding validation code for -ve values to `raise value error`.

Attaching [gist](https://colab.research.google.com/gist/SuryanarayanaY/650edf2e880caa947674f813a41ef5a8/61158_-code-fix-ve-values.ipynb) for the same with solution for reference.

Fixex #61158

SuryanarayanaY committed:
validate clip_norm argument in clip_by_norm API

The API clip_by_norm have argument clip_norm which accepts  0-D (scalar) `Tensor` > 0 . But if we pass -ve value for this argument then its not raising intended error and converting the input tensor into Negative which IMO is wrong. Hence I am adding validation code for -ve values to raise value error.

cantonios commented on tensorflow/python/ops/clip_ops.py:
```diff
@@ -212,6 +212,8 @@ def clip_by_norm(t, clip_norm, axes=None, name=None):
     values = ops.convert_to_tensor(
         t.values if isinstance(t, indexed_slices.IndexedSlices) else t,
         name="t")
+    if clip_norm < 0:
```
This won't work if `clip_norm` is a tensor.

I think your best option is to set
```
clip_norm = math_ops.maximum(clip_norm, 0)
```
to force it to be positive.

SuryanarayanaY replied:
@cantonios 
Thanks for the inputs. Actually I have tested with a Tensor also by passing `clip_norm=tf.constant(-10.0)` and the proposed solution still working and same can be referred in the gist attached in above [comment](https://github.com/tensorflow/tensorflow/pull/61242#issue-1798546174).

I have also tested with `tf.function` and the proposed solution still working and raising intended error.Please refer [gist2](https://colab.research.google.com/gist/SuryanarayanaY/40fd4154a5835c65c5493407cd2f9ef8/61158_errors-for-code-fix-2.ipynb).

for example below code outputs bool Tensor and hence it seems working for this case:

```
clip_norm=tf.constant(-10.0)
print(clip_norm <0)

#output:tf.Tensor(True, shape=(), dtype=bool)
```
I am not sure though which cases it may fail.

cantonios replied:
That's because you're giving it constant tensors, so in the `tf.function` can extract the constant value during tracing.  If it was a variable tensor, it would fail.

For example, the following works fine with the original `tf.clip_by_norm`:
```python
@tf.function
def tf_clip_by_norm(t, clip_norm, axes=[1]):
  return tf.clip_by_norm(t, clip_norm=clip_norm, axes=[1])

# Current tf.clip_by_norm:
clip = tf_clip_by_norm.get_concrete_function(tf.TensorSpec(shape=[3, 3], dtype=tf.float32), tf.TensorSpec(shape=[], dtype=tf.float32), tf.TensorSpec(shape=[], dtype=tf.int32))
print(clip(tf.ones([3, 3]), tf.constant(0.5), tf.constant(1)))
```

But with your modification, it incorrectly throws the `ValueError` immediately when trying to get the concrete function:
```python
clip = clip_by_norm.get_concrete_function(tf.TensorSpec(shape=[3, 3], dtype=tf.float32), tf.TensorSpec(shape=[], dtype=tf.float32), tf.TensorSpec(shape=[], dtype=tf.int32))
```
```python
ValueError: in user code:

    File "<ipython-input-9-abfa95055a32>", line 22, in clip_by_norm  *
        raise ValueError('clip_norm should be a 0-D (scalar) Tensor > 0')

    ValueError: clip_norm should be a 0-D (scalar) Tensor > 0
```
even though we haven't actually passed in any values.

SuryanarayanaY replied:
@cantonios ,

Thanks for clarification. Then I will go with converting -ve values to 0 as suggested using:

`clip_norm = math_ops.maximum(clip_norm, 0)
`

Also I would like to add a note of same in the Documentation to make the users aware of this.

Thanks

cantonios replied:
Yes, please update the documentation.

SuryanarayanaY committed:
Update clip_ops.py

Changes done to convert clip_norm to Zero and also added a note in documentation.

cantonios commented on tensorflow/python/ops/clip_ops.py:
```diff
@@ -194,7 +194,9 @@ def clip_by_norm(t, clip_norm, axes=None, name=None):
     clip_norm: A 0-D (scalar) `Tensor` > 0. A maximum clipping value, also
-      floating point
+      floating point.
+      Note: If clip_norm given as a negative value then it will be converted
```
- clip_norm given -> clip_norm is given
- Zero -> zero
- internally and exception will not be raised here -> without raising an exception

Alternatively, just say:
> If a negative clip_norm is provided, it will be treated as zero.

SuryanarayanaY committed:
Update clip_ops.py

Updated clip_norm note as requested. Thanks!

gbaned commented:
Here are the internal errors, @SuryanarayanaY can you please verify ? Thank you!

Traceback (most recent call last):
  File "/tensorflow/python/ops/math_ops.py", line 1486, in binary_op_wrapper
    out = r_op(x)
  File "/tensorflow/python/util/traceback_utils.py", line 141, in error_handler
    return fn(*args, **kwargs)
  File "/tensorflow/python/ops/math_ops.py", line 1509, in r_binary_op_wrapper
    y, x = maybe_promote_tensors(y, x, force_same_dtype=True)
  File "/tensorflow/python/ops/math_ops.py", line 1441, in maybe_promote_tensors
    ops.convert_to_tensor(tensor, dtype, name="x"))
  File "/tensorflow/python/profiler/trace.py", line 183, in wrapped
    return func(*args, **kwargs)
  File "/tensorflow/python/framework/ops.py", line 701, in convert_to_tensor
    return tensor_conversion_registry.convert(
  File "/tensorflow/python/framework/tensor_conversion_registry.py", line 209, in convert
    return overload(dtype, name)  #  pylint: disable=not-callable
  File "/tensorflow/python/framework/tensor.py", line 762, in __tf_tensor__
    raise ValueError(
ValueError: x: Tensor conversion requested dtype float32 for Tensor with dtype float64: <tf.Tensor 'embedding_lookup:0' shape=() dtype=float64>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tensorflow/python/framework/test_util.py", line 1634, in decorated
    return f(self, *args, **kwargs)
  File "/tensorflow/python/kernel_tests/nn_ops/embedding_ops_test.py", line 584, in testHigherRankMaxNorm
    simple = embedding_ops.embedding_lookup(params, ids, max_norm=1.0)
  File "/tensorflow/python/util/traceback_utils.py", line 141, in error_handler
    return fn(*args, **kwargs)
  File "/tensorflow/python/util/dispatch.py", line 1260, in op_dispatch_handler
    return dispatch_target(*args, **kwargs)
  File "/tensorflow/python/ops/embedding_ops.py", line 326, in embedding_lookup
    return _embedding_lookup_and_transform(
  File "/tensorflow/python/ops/embedding_ops.py", line 144, in _embedding_lookup_and_transform
    result = _clip(
  File "/tensorflow/python/ops/embedding_ops.py", line 73, in _clip
    return clip_ops.clip_by_norm(
  File "/tensorflow/python/util/traceback_utils.py", line 141, in error_handler
    return fn(*args, **kwargs)
  File "/tensorflow/python/util/dispatch.py", line 1260, in op_dispatch_handler
    return dispatch_target(*args, **kwargs)
  File "/tensorflow/python/ops/clip_ops.py", line 223, in clip_by_norm
    intermediate = values * clip_norm
  File "/tensorflow/python/util/traceback_utils.py", line 141, in error_handler
    return fn(*args, **kwargs)
  File "/tensorflow/python/ops/math_ops.py", line 1491, in binary_op_wrapper
    raise e
  File "/tensorflow/python/ops/math_ops.py", line 1475, in binary_op_wrapper
    return func(x, y, name=name)
  File "/tensorflow/python/ops/math_ops.py", line 1856, in _mul_dispatch
    return multiply(x, y, name=name)
  File "/tensorflow/python/util/traceback_utils.py", line 141, in error_handler
    return fn(*args, **kwargs)
  File "/tensorflow/python/util/dispatch.py", line 1260, in op_dispatch_handler
    return dispatch_target(*args, **kwargs)
  File "/tensorflow/python/ops/math_ops.py", line 529, in multiply
    return gen_math_ops.mul(x, y, name)
  File "/tensorflow/python/ops/gen_math_ops.py", line 6590, in mul
    _, _, _op, _outputs = _op_def_library._apply_op_helper(
  File "/tensorflow/python/framework/op_def_library.py", line 778, in _apply_op_helper
    _ExtractInputsAndAttrs(op_type_name, op_def, allowed_list_attr_map,
  File "/tensorflow/python/framework/op_def_library.py", line 589, in _ExtractInputsAndAttrs
    raise TypeError(
TypeError: Input 'y' of 'Mul' Op has type float32 that does not match type float64 of argument 'x'.

gbaned commented:
Hi @SuryanarayanaY Any update on this PR? Please. Thank you!

SuryanarayanaY commented:
I have gone through the error stack and root cause is as below.

> File "/tensorflow/python/kernel_tests/nn_ops/embedding_ops_test.py", line 584, in testHigherRankMaxNorm
> simple = embedding_ops.embedding_lookup(params, ids, max_norm=1.0)

In the above test the `params` is `float64` and max_norm=1.0 which are then passed to `clip_by_norm()` as `t` and `clip_norm` respectively. 

> 
> File "/tensorflow/python/ops/clip_ops.py", line 223, in clip_by_norm
> intermediate = values * clip_norm

Then in the above line of code `t` is converted to `values ` as `float64` tensor and `clip_norm=1.0`
```
values = ops.convert_to_tensor(
        t.values if isinstance(t, indexed_slices.IndexedSlices) else t,
        name="t")
```

With the proposed amendment `clip_norm = math_ops.maximum(clip_norm, 0)`, the `clip_norm` becoming float32 tensor and this is causing the error and test fail. Without modification it remains as 1.0 and hence `values * clip_norm` works fine here.

One observation is the args `t` and `clip_norm` of API `tf.clip_by_norm` both the args should be of same dtype else exception will be raised which is also not documented.

My proposed solution is:

```
clip_norm = tf.cast(math_ops.maximum(clip_norm, 0),dtype = values.dtype)
```

which can validate -ve inputs, resolves the issue mentioned above regarding dtype of `t` and `clip_norm` and it also eliminates testfail.

I have added a [gist](https://colab.sandbox.google.com/gist/SuryanarayanaY/f01ac87fc5e19aebca7be360adef2ae2/clip_by_norm-pr-61242.ipynb) for above exercise for reference.

@cantonios , Needs your review and comments for proceeding.

gbaned commented:
Hi @cantonios, Can you please review this PR ? Thank you!

gbaned commented:
Hi @cantonios, Can you please review this PR ? Thank you!

SuryanarayanaY committed:
validate clip_norm argument in clip_ops.py

Earlier changes causing few test fails.Hence changed the validation code and applied casting to convert dtype to dtype = values.dtype which may avoid test fails.

cantonios commented on tensorflow/python/ops/clip_ops.py:
```diff
@@ -212,6 +213,8 @@ def clip_by_norm(t, clip_norm, axes=None, name=None):
     values = ops.convert_to_tensor(
         t.values if isinstance(t, indexed_slices.IndexedSlices) else t,
         name="t")
+    clip_norm = tf.cast(math_ops.maximum(clip_norm, 0),dtype = values.dtype)
```
You won't be able to use `tf.cast`.  I think it's in `math_ops`.

SuryanarayanaY replied:
Apologies.Yes, It should be `math_ops.cast`.

gbaned commented:
Hi @SuryanarayanaY Can you please check @cantonios's comments and keep us posted ? Thank you!

SuryanarayanaY committed:
Update clip_ops.py

change tf.cast to math_ops.cast
